### ISSUE TITLE
# About metric-scale depth estimation and point cloud reconstruction

### ISSUE THREAD

**Issue by bao-tg** [ByteDance-Seed/Depth-Anything-3#244]
## Problems
I have tried to use both DA3NESTED-GIANT-LARGE, DA3NESTED-GIANT-LARGE-1.1, DA3METRIC-LARGE on my data. However, the results aren't good as expected (I run inference to get npz file -> run my script below to get ply file).

For instance, below are my reconstructed point clouds using the model DA3NESTED-GIANT-LARGE, DA3NESTED-GIANT-LARGE-1.1, DA3METRIC-LARGE, note that the true scale is 9cm x 16.2cm. However, the result that I get was ~ 0.166 x 0.278 (In theory, if the depth is correctly metric-scale, and given camera's intrinsic, in pixels, I can back-project 2D depth image into metric-scale 3D point cloud). I will put my implementation below.

If my implementation is correct, I guess this is probably the limitation of neural-based model on metric-scale depth estimation, which is failing when being applied to out-of-domain data. And perhaps, fine-tuning is the only way to mitigate this.

I really hope that I made the mistake somewhere, anyone met a problem similar this? can I get your recommendation on how to get correct metric-scale 3D point clouds.

### DA3NESTED-GIANT-LARGE

Note that the outputs by DA3NESTED-GIANT-LARGE and LARGE1.1 have already been in meters.

<img width="581" height="533" alt="Image" src="https://github.com/user-attachments/assets/e8e3b665-520b-4e60-b775-5c61cc7ecc70" />

### DA3NESTED-GIANT-LARGE-1.1

<img width="581" height="533" alt="Image" src="https://github.com/user-attachments/assets/2c88d0ed-c7b9-4f17-9a30-a2504c709ff9" />

### DA3METRIC-LARGE

This model is even worse compared to two aforementioned models, since it predicted the wrong scale {width x height}.

<img width="581" height="533" alt="Image" src="https://github.com/user-attachments/assets/ff460b0e-5dfe-474e-ac32-ab51af41c09b" />

## MY IMPLEMENTATION TO TRANSFER RESULT.NPZ TO PLY FILE

```
import numpy as np
import cv2
import os

# ─────────────────────────────────────────────
# 1. LOAD
# ─────────────────────────────────────────────
# data = np.load()
depth = data['depth'].squeeze(0)       # (H_depth, W_depth)

# rgb = cv2.imread()
rgb = cv2.cvtColor(rgb, cv2.COLOR_BGR2RGB)

# ─────────────────────────────────────────────
# 2. INTRINSICS — original image resolution
# ─────────────────────────────────────────────
H_orig, W_orig = 1080, 1920
fx          = 1360.637451
fy          = 1359.588745
cx          = 967.345520
cy          = 537.803223

# ─────────────────────────────────────────────
# 3. SCALE INTRINSICS TO DEPTH RESOLUTION
# ─────────────────────────────────────────────
H_depth, W_depth = depth.shape
print(f"Depth resolution : {H_depth} x {W_depth}")
print(f"RGB resolution   : {rgb.shape[0]} x {rgb.shape[1]}")

fx_s = fx * (W_depth / W_orig)
fy_s = fy * (H_depth / H_orig)
cx_s = cx * (W_depth / W_orig)
cy_s = cy * (H_depth / H_orig)
print(f"Scaled intrinsics: fx={fx_s:.3f}, fy={fy_s:.3f}, cx={cx_s:.3f}, cy={cy_s:.3f}")

# ─────────────────────────────────────────────
# 4. RESIZE RGB TO MATCH DEPTH RESOLUTION
# ─────────────────────────────────────────────
rgb_resized = cv2.resize(rgb, (W_depth, H_depth), interpolation=cv2.INTER_LINEAR)

# ─────────────────────────────────────────────
# 5. BUILD PIXEL GRID
# ─────────────────────────────────────────────
u = np.arange(W_depth)
v = np.arange(H_depth)
uu, vv = np.meshgrid(u, v)             # both (H_depth, W_depth)

# ─────────────────────────────────────────────
# 6. UNPROJECT TO 3D
# ─────────────────────────────────────────────
Z = depth
X = (uu - cx_s) * Z / fx_s
Y = (vv - cy_s) * Z / fy_s

# ─────────────────────────────────────────────
# 7. FILTER INVALID DEPTHS
# ─────────────────────────────────────────────
valid = (Z > 0) & np.isfinite(Z)
X = X[valid]
Y = Y[valid]
Z = Z[valid]
colors = rgb_resized[valid].astype(np.uint8)
points = np.stack([X, Y, Z], axis=-1)  # (N, 3)
print(f"Total valid points: {len(points)}")
print(f"Depth range: {Z.min():.3f} ~ {Z.max():.3f} m")

# ─────────────────────────────────────────────
# 8. SAVE AS PLY WITH COLOR
# ─────────────────────────────────────────────
def save_ply_rgb(filepath, points, colors):
    os.makedirs(os.path.dirname(filepath), exist_ok=True)
    header = (
        "ply\n"
        "format ascii 1.0\n"
        f"element vertex {len(points)}\n"
        "property float x\n"
        "property float y\n"
        "property float z\n"
        "property uchar red\n"
        "property uchar green\n"
        "property uchar blue\n"
        "end_header\n"
    )
    with open(filepath, 'w') as f:
        f.write(header)
        for p, c in zip(points, colors):
            f.write(f"{p[0]:.6f} {p[1]:.6f} {p[2]:.6f} {c[0]} {c[1]} {c[2]}\n")
    print(f"Saved → {filepath}")

save_ply_rgb(PATH, points, colors)

```

**Comment by sourcesync** [comment]
@bao-tg Did you apply the focal scaling to DA3METRIC-LARGE?

**Comment by bao-tg** [comment]
> [@bao-tg](https://github.com/bao-tg) Did you apply the focal scaling to DA3METRIC-LARGE?

```bash
Monocular Metric Depth: To obtain metric depth in meters from DA3METRIC-LARGE, 
use metric_depth = focal * net_output / 300., where focal is the focal length in pixels 
(typically the average of fx and fy from the camera intrinsic matrix K). Note that the output from 
DA3NESTED-GIANT-LARGE is already in meters.
```
For the model DA3NESTED-GIANT-LARGE and LARGE1.1, I didn't scale as the above note from dev team. 

For the model DA3METRIC-LARGE, even if I scale it by `focal/300`, it still gives a wrongly scale output.

**Comment by sourcesync** [comment]
Thanks @bao-tg .  In my case, when I scale DA3METRIC-LARGE by the intrinsics produced from DA3NESTED-GIANT-LARGE-1.1, the estimates are closer to ground truth.  But, in my case, depths are in the 0.5 m range.  It looks like you are much closer ?

**Comment by bao-tg** [comment]
> Thanks @bao-tg .  In my case, when I scale DA3METRIC-LARGE by the intrinsics produced from DA3NESTED-GIANT-LARGE-1.1, the estimates are closer to ground truth.  But, in my case, depths are in the 0.5 m range.  It looks like you are much closer ?

Thanks for your reply. Instead of using camera's intrinsics produced by the model, I use exact intrinsics from my camera (as you can see in my provided code above), I believed that it must be better than predicted intrinsics.

I'm not sure if I understand your point, even if depths' range are closer than 0.5m, it doesn't mean the model's output is correct, I use camera intrinsics to visualize the depth in 3D. If the depth is perfectly correct, then my point cloud need to be perfectly correct too (it's maths). In my case, the pcl is wrongly scaled -> the depth is wrongly scaled too.

**Comment by sourcesync** [comment]
@bao-tg  My point about the range...my ground truth objects are at the 0.5 range, but it looks like yours are closer.  Could it be the model does not do well on closer objects in this range?  That was my point and I don't have a good answer, just thinking out loud.

As for the intrinsics, maybe one thing to try is a model that produces intrinsics to see how far it is from your "exact" one.  

My apologies if none of these ideas are useful, I'm just getting started with DA3 and VGGT depth estimation models myself.  I'm also trying the DA2 metric models.

**Comment by bao-tg** [comment]
> [@bao-tg](https://github.com/bao-tg) My point about the range...my ground truth objects are at the 0.5 range, but it looks like yours are closer. Could it be the model does not do well on closer objects in this range? That was my point and I don't have a good answer, just thinking out loud.
> 
> As for the intrinsics, maybe one thing to try is a model that produces intrinsics to see how far it is from your "exact" one.
> 
> My apologies if none of these ideas are useful, I'm just getting started with DA3 and VGGT depth estimation models myself. I'm also trying the DA2 metric models.

It's okay, I appreciate your recommendation. Actually, the predicted intrinsics and the actual camera's intrinsics are the same (If you rescale the predicted intrinsics to match the image's original size, I found that on one of my exps, note that the model crops the size of original images and train-test on these cropped images). 

Maybe the problem is as you said, "the model doesn't do well on closer objects".

**Comment by sourcesync** [comment]
@bao-tg I did another set of experiments - one with objects at 0.25m and one with objects at 0.73m.  The relative errors for 0.73 m are under 10%, while the relative error is well over 100% for the closer objects!  I tried DA3 (metric only), DA3 (metric only), VGGT (not sure if its metric), and GenAI models (Gemma3, Gemma4, Google-Pro-3, Claude-Sonnet-4.5).  Gemma4 performs the best at the closer distance but rel error is still high above 30%.  I see that DA2 has a set fine-tuning scripts and I'm curious if applying it to a dataset under 0.5m would help.  If I end up doing something like this, I'll let you know.
